File: node/tests/riskService.test.js
import { RiskService } from '../src/services/riskService.js';

describe('RiskService', () => {
    const service = new RiskService();

    test('Low risk renter qualification', () => {
        const profile = {
            id: 'r1',
            annual_income: 100000,
            credit_score: 780,
            monthly_rent: 2000
        };
        const result = service.assessRenter(profile);
        expect(result.qualified).toBe(true);
        expect(result.risk_score).toBeLessThanOrEqual(30);
        expect(result.recommendation).toBe('Approve');
    });

    test('High risk - low credit score rejection', () => {
        const profile = {
            id: 'r2',
            annual_income: 50000,
            credit_score: 550,
            monthly_rent: 2500
        };
        const result = service.assessRenter(profile);
        expect(result.qualified).toBe(false);
        expect(result.risk_score).toBeGreaterThanOrEqual(70);
        expect(result.recommendation).toBe('Reject - High Risk');
    });

    test('Moderate risk - high rent ratio qualification', () => {
        const profile = {
            id: 'r3',
            annual_income: 60000,
            credit_score: 680,
            monthly_rent: 2500
        };
        const result = service.assessRenter(profile);
        expect(result.qualified).toBe(true);
        expect(result.risk_score).toBeGreaterThanOrEqual(40);
        expect(result.risk_score).toBeLessThanOrEqual(65);
        expect(result.recommendation).toBe('Approve with Co-signer');
    });
});

File: internal/service/risk_service_test.go
package service

import (
	"testing"

	"github.com/securerent/risk-assessment-service/internal/model"
)

func TestAssessRenter(t *testing.T) {
	s := NewRiskService()

	tests := []struct {
		name           string
		profile        model.RenterProfile
		wantQualified  bool
		minRiskScore   int
		maxRiskScore   int
	}{
		{
			name: "Low risk renter",
			profile: model.RenterProfile{
				AnnualIncome: 100000,
				CreditScore:  780,
				MonthlyRent:  2000,
			},
			wantQualified: true,
			minRiskScore:  0,
			maxRiskScore:  30,
		},
		{
			name: "High risk - low credit",
			profile: model.RenterProfile{
				AnnualIncome: 50000,
				CreditScore:  550,
				MonthlyRent:  2500,
			},
			wantQualified: false,
			minRiskScore:  70,
			maxRiskScore:  100,
		},
		{
			name: "Moderate risk - high rent ratio",
			profile: model.RenterProfile{
				AnnualIncome: 60000,
				CreditScore:  680,
				MonthlyRent:  2500,
			},
			wantQualified: true,
			minRiskScore:  40,
			maxRiskScore:  65,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := s.AssessRenter(tt.profile)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got.Qualified != tt.wantQualified {
				t.Errorf("Qualified = %v, want %v", got.Qualified, tt.wantQualified)
			}
			if got.RiskScore < tt.minRiskScore || got.RiskScore > tt.maxRiskScore {
				t.Errorf("RiskScore = %d, want between %d and %d", got.RiskScore, tt.minRiskScore, tt.maxRiskScore)
			}
		}
	}
}
